Libvirtd ConfigSection Subclass

jirihnidek opened the pull request:
* Implement subclass of VirtConfigSection used for validation
  options and combination of options specific for libvirtd
  virtualization backend
* Moved some validation from VirtConfigSection to
  LibvirtdConfigSection. Ported remaining libvirtd validation
  from `chechOptions` of old `Config` class.
* Improved method `class_for_type` to search recursively for
  subclasses in subclasses of ConfigSection.
* Implemented `__new__()` method fot VirtConfigSection to
  return instances of VirtConfigSection subclasses with
  required `virt_type`, but it is not used anywhere ATM.
* Added several unit tests

jirihnidek commented:
![](https://github.trello.services/images/mini-trello-icon.png) [[13] [CONFIG_EPIC] Create Libvirtd Config Subclass](https://trello.com/c/RHPohoJF/530-13-configepic-create-libvirtd-config-subclass)

cnsnyder commented on virtwho/config.py:
```diff
@@ -1264,6 +1288,117 @@ def _validate(self):
+    def __init__(self, section_name, wrapper):
+        super(LibvirtdConfigSection, self).__init__(section_name, wrapper)
+
+    def _validate_server(self):
```
I would suggest that we do whatever sort of checking is necessary for the scheme etc being used throughout libvirtd here. (See Libvirtd._get_url)

In this way we can show what sort of problems might occur long before we try to use the values.

jirihnidek replied:
I want to clarify some things.

Your suggestion seems logical, but I don't see any validation in `Libvirtd._get_url()` in terms of exit this method unless condition is met. What I see there is logging of warnings, when scheme is not complete and some default value is injected to scheme.

I don't recommend moving content of  `Libvirtd._get_url()` to `LIbvirtdConfigSection._validate_server()`, because purpose of all `_validate*()` method is validation of configuration files/CLI/ENV and not changing of loaded configuration options, right?

If we want to change loaded options in `ConfigSection` and subclasses of `ConfigSection`, then I'm not against, but we should do it in another method.

cnsnyder replied:
The purpose of _validate*() methods is validating the config values from whatever source they may be as well as to make any changes necessary for that parameter to be usable (so long as there is a reasonable way to do that with the given parameter).

There are more things I'd like to be able to verify (like the connectivity to the source/destinations for the given configuration) but such things can wait.

cnsnyder reviewed (changes requested):
Minor requested changes. Overall, looks good!!

Thank you for the contribution. Please address the comments and I'll get back to you later!

Cheers,
Chris

cnsnyder reviewed (changes requested):
In addition, any chance we could move the libvirtdconfigsection to the libvirtd.py file?


Running 'virt-who -o -d' with the following configuration fails as described below:

## Config
```ini
[LOCAL_LIBVIRT]
type=libvirt
server=qemu:///system
owner=admin
env=admin
rhsm_hostname=localhost
rhsm_port=8443
rhsm_username=admin
rhsm_password=admin
rhsm_prefix=/candlepin
#filter_hosts=d5b15581-5440-11cb-930e-b5dfde6103bc
hypervisor_id=hostname
```

## Error

"""
2017-09-07 16:48:55,812 [virtwho.config DEBUG] MainProcess(13780):MainThread @config.py:all_drop_dir_config_sections:1619 - There are files in '/etc/virt-who.d/' not ending in '*.conf' is this intentional?
__init__() takes exactly 4 arguments (3 given)
Traceback (most recent call last):
  File "/home/csnyder/devel/virt-who-upstream/virtwho/__main__.py", line 11, in main
    res = virtwho.main.main()
  File "virtwho/main.py", line 114, in main
    logger, options = parse_options()
  File "virtwho/parser.py", line 492, in parse_options
    effective_config = init_config(env_options, cli_options)
  File "virtwho/config.py", line 1746, in init_config
    new_section = ConfigSection.from_dict(new_section, section, effective_config)
  File "virtwho/config.py", line 986, in from_dict
    section = cls.class_for_type(virt_type)(section_name, wrapper)
TypeError: __init__() takes exactly 4 arguments (3 given)
No handlers could be found for logger "virtwho.main"
"""

I think all that is necessary is a change to that line to pass in the virt_type.

After such a fix there, libvirt is still unable to connect to my local system as specified. This works in master and I believe it should work here as well.

jirihnidek commented:
I fixed the bug and I added unit test for this case. I'm able to connect to `libvirtd` and report guests to candlepin server after this bug fix.

cnsnyder commented on virtwho/virt/libvirtd/libvirtd.py:
```diff
@@ -115,13 +115,14 @@ def isHypervisor(self):
 
     def _get_url(self):
         if self.config.get('server', None):
+            # FIXME: Following code can be removed, when ConfigSection will be used
```
We should remove this now as the LibvirtdConfigSection should be in use after this PR is merged.

cnsnyder reviewed (changes requested):
I would request a change to libvirtd.py:Libvirtd._connect:
```python
        url = self.config.get('server', None)
```

And that we remove the old _get_url code as the libvirtdConfigSection should be in use after this is merged.

cnsnyder commented:
This still does not work with the following configuration
## Configuration
```ini
[LOCAL_LIBVIRT]
type=libvirt
server=qemu:///system
owner=admin
env=admin
rhsm_hostname=localhost
rhsm_port=8443
rhsm_username=admin
rhsm_password=admin
rhsm_prefix=/candlepin
```

## Output
```
2017-09-26 17:55:25,144 [virtwho.config DEBUG] MainProcess(22870):MainThread @config.py:all_drop_dir_config_sections:1621 - There are files in '/etc/virt-who.d/' not ending in '*.conf' is this intentional?
<class 'virtwho.config.LibvirtdConfigSection'>
2017-09-26 17:55:25,145 [virtwho.config WARNING] MainProcess(22870):MainThread @config.py:init_config:1761 - Value for log_per_config not set in: global, using default: False
2017-09-26 17:55:25,145 [virtwho.config WARNING] MainProcess(22870):MainThread @config.py:init_config:1761 - Value for print not set in: global, using default: False
2017-09-26 17:55:25,145 [virtwho.config WARNING] MainProcess(22870):MainThread @config.py:init_config:1761 - Value for reporter_id not set in: global, using default: csnyder-laptop-ab7e6eec07bd4f69a53afc4970d7162e
2017-09-26 17:55:25,145 [virtwho.config WARNING] MainProcess(22870):MainThread @config.py:init_config:1761 - Value for log_dir not set in: global, using default: /var/log/rhsm
2017-09-26 17:55:25,145 [virtwho.config WARNING] MainProcess(22870):MainThread @config.py:init_config:1761 - Value for background not set in: global, using default: False
2017-09-26 17:55:25,146 [virtwho.config WARNING] MainProcess(22870):MainThread @config.py:init_config:1761 - Value for configs not set in: global, using default: []
2017-09-26 17:55:25,146 [virtwho.config WARNING] MainProcess(22870):MainThread @config.py:init_config:1761 - Value for log_file not set in: global, using default: rhsm.log
2017-09-26 17:55:25,146 [virtwho.init WARNING] MainProcess(22870):MainThread @parser.py:parse_options:502 - Unable to check cli argument consistency, no known destination provided
2017-09-26 17:55:25,146 [virtwho.init WARNING] MainProcess(22870):MainThread @parser.py:parse_options:502 - Value for simplified_vim not set in: LOCAL_LIBVIRT, using default: True
2017-09-26 17:55:25,146 [virtwho.init WARNING] MainProcess(22870):MainThread @parser.py:parse_options:502 - Value for filter_host_uuids not set in: LOCAL_LIBVIRT, using default: None
2017-09-26 17:55:25,146 [virtwho.init WARNING] MainProcess(22870):MainThread @parser.py:parse_options:502 - Value for hypervisor_id not set in: LOCAL_LIBVIRT, using default: uuid
2017-09-26 17:55:25,146 [virtwho.init WARNING] MainProcess(22870):MainThread @parser.py:parse_options:502 - Value for is_hypervisor not set in: LOCAL_LIBVIRT, using default: True
2017-09-26 17:55:25,146 [virtwho.init WARNING] MainProcess(22870):MainThread @parser.py:parse_options:502 - Value for sm_type not set in: LOCAL_LIBVIRT, using default: sam
2017-09-26 17:55:25,146 [virtwho.init WARNING] MainProcess(22870):MainThread @parser.py:parse_options:502 - Value for filter_hosts not set in: LOCAL_LIBVIRT, using default: None
2017-09-26 17:55:25,146 [virtwho.init WARNING] MainProcess(22870):MainThread @parser.py:parse_options:502 - Value for filter_host_parents not set in: LOCAL_LIBVIRT, using default: None
2017-09-26 17:55:25,146 [virtwho.init WARNING] MainProcess(22870):MainThread @parser.py:parse_options:502 - Value for exclude_host_uuids not set in: LOCAL_LIBVIRT, using default: None
2017-09-26 17:55:25,146 [virtwho.init WARNING] MainProcess(22870):MainThread @parser.py:parse_options:502 - Value for exclude_hosts not set in: LOCAL_LIBVIRT, using default: None
2017-09-26 17:55:25,146 [virtwho.init WARNING] MainProcess(22870):MainThread @parser.py:parse_options:502 - Value for exclude_host_parents not set in: LOCAL_LIBVIRT, using default: None
2017-09-26 17:55:25,147 [virtwho.init WARNING] MainProcess(22870):MainThread @parser.py:parse_options:502 - Ignoring unknown configuration option "rhsm_hostname"
2017-09-26 17:55:25,147 [virtwho.init WARNING] MainProcess(22870):MainThread @parser.py:parse_options:502 - Ignoring unknown configuration option "rhsm_port"
2017-09-26 17:55:25,147 [virtwho.init WARNING] MainProcess(22870):MainThread @parser.py:parse_options:502 - Ignoring unknown configuration option "rhsm_prefix"
2017-09-26 17:55:25,147 [virtwho.init WARNING] MainProcess(22870):MainThread @parser.py:parse_options:502 - No values provided in: LOCAL_LIBVIRT
2017-09-26 17:55:25,147 [virtwho.init WARNING] MainProcess(22870):MainThread @parser.py:parse_options:502 - Value for log_per_config not set in: global, using default: False
2017-09-26 17:55:25,147 [virtwho.init WARNING] MainProcess(22870):MainThread @parser.py:parse_options:502 - Value for print not set in: global, using default: False
2017-09-26 17:55:25,147 [virtwho.init WARNING] MainProcess(22870):MainThread @parser.py:parse_options:502 - Value for reporter_id not set in: global, using default: csnyder-laptop-ab7e6eec07bd4f69a53afc4970d7162e
2017-09-26 17:55:25,147 [virtwho.init WARNING] MainProcess(22870):MainThread @parser.py:parse_options:502 - Value for log_dir not set in: global, using default: /var/log/rhsm
2017-09-26 17:55:25,147 [virtwho.init WARNING] MainProcess(22870):MainThread @parser.py:parse_options:502 - Value for background not set in: global, using default: False
2017-09-26 17:55:25,147 [virtwho.init WARNING] MainProcess(22870):MainThread @parser.py:parse_options:502 - Value for configs not set in: global, using default: []
2017-09-26 17:55:25,147 [virtwho.init WARNING] MainProcess(22870):MainThread @parser.py:parse_options:502 - Value for log_file not set in: global, using default: rhsm.log
2017-09-26 17:55:25,147 [virtwho.init WARNING] MainProcess(22870):MainThread @parser.py:parse_options:502 - No values provided in: global
2017-09-26 17:55:25,147 [virtwho.init WARNING] MainProcess(22870):MainThread @parser.py:parse_options:502 - No values provided in: global
2017-09-26 17:55:25,147 [virtwho.init DEBUG] MainProcess(22870):MainThread @executor.py:__init__:45 - Using config named 'LOCAL_LIBVIRT'
2017-09-26 17:55:25,148 [virtwho.init INFO] MainProcess(22870):MainThread @main.py:main:158 - Using configuration "LOCAL_LIBVIRT" ("libvirt" mode)
2017-09-26 17:55:25,148 [virtwho.init INFO] MainProcess(22870):MainThread @main.py:main:160 - Using reporter_id='csnyder-laptop-ab7e6eec07bd4f69a53afc4970d7162e'
2017-09-26 17:55:25,224 [rhsm.https DEBUG] MainProcess(22870):MainThread @https.py:<module>:56 - Using standard libs to provide httplib and ssl
2017-09-26 17:55:25,229 [virtwho.LOCAL_LIBVIRT DEBUG] MainProcess(22870):Thread-2 @virt.py:run:374 - Thread 'LOCAL_LIBVIRT' started
2017-09-26 17:55:25,229 [virtwho.LOCAL_LIBVIRT INFO] MainProcess(22870):Thread-2 @libvirtd.py:_connect:126 - Using libvirt url: *qemu://None/system?no_tty=1*
2017-09-26 17:55:25,230 [virtwho.LOCAL_LIBVIRT ERROR] MainProcess(22870):Thread-2 @libvirtd.py:_connect:134 - Error in libvirt backend
Traceback (most recent call last):
  File "virtwho/virt/libvirtd/libvirtd.py", line 132, in _connect
    v = libvirt.openReadOnly(url)
  File "/usr/lib64/python2.7/site-packages/libvirt.py", line 268, in openReadOnly
    if ret is None:raise libvirtError('virConnectOpenReadOnly() failed')
libvirtError: Cannot read CA certificate '/etc/pki/CA/cacert.pem': No such file or directory
2017-09-26 17:55:25,231 [virtwho.LOCAL_LIBVIRT ERROR] MainProcess(22870):Thread-2 @virt.py:run:383 - Thread 'LOCAL_LIBVIRT' fails with error: Cannot read CA certificate '/etc/pki/CA/cacert.pem': No such file or directory
```

Notice above the url that the libvirt  backend attempts to use is "qemu://None/system?no_tty=1"

I believe this should be fixed before merge.

cnsnyder commented on virtwho/config.py:
```diff
@@ -1655,12 +1663,13 @@ def _check_effective_config_validity(effective_config):
             if name == VW_GLOBAL:
                 # Always keep the global section
                 continue
+            # FIXME: why is env/cli automatically dropped?
```
It should not be so long as it is valid

cnsnyder commented on virtwho/config.py:
```diff
@@ -1264,6 +1295,164 @@ def _validate(self):
         self.validation_messages.extend(validation_messages)
 
 
+class LibvirtdConfigSection(VirtConfigSection):
```
Please move this to the libvirtd.py file.

cnsnyder reviewed (changes requested):
Last requested change. Please move the libvirt config section to the libvirtd file.

Once that is done we'll have a great example of how to implement validation for a particular backend (so we can continue on the other backends).

cnsnyder commented:
@jirihnidek Please rebase from config_refactor. Then this will be ready for merge!!

jirihnidek committed:
Libvirtd ConfigSection Subclass

* Implement subclass of VirtConfigSection used for validation
  options and combination of options specific for libvirtd
  virtualization backend
* LibvirtdConfigSection is implemented in libvirtd.py not in
  config.py
* Moved some validation from VirtConfigSection to
  LibvirtdConfigSection. Ported remaining libvirtd validation
  from `chechOptions` of old `Config` class.
* Improved method `class_for_type` to search recursively for
  subclasses in subclasses of ConfigSection.
* Implemented `__new__()` method fot VirtConfigSection to
  return instances of VirtConfigSection subclasses with
  required `virt_type`, but it is not used anywhere ATM.
* Implemented validation of libvirtd URL. It was ported from
  Libvirtd._get_url and found one bug there. It will be fixed
  in another commit
* Validation of URL is part of `_validate_server()`
* This function also changes server URL, when it is needed
* Added several unit tests fot testing validation
* Fixed `__new__()` method of VirtConfigSection and added
  unit test for this feature.
* Removed function _get_url() and use benefit of new class
  LibvirtdConfigSection
* Fix errors in libvirtd validation
* Fix some other errors
   - filters
   - reading config file using -c option
* Added several other unit tests

jirihnidek committed:
Merge branch 'config_refactor' into jhnidek/libvirtd-config_section

Conflicts:
	virtwho/config.py

cnsnyder commented:
Merging as this looks good overall. The broken tests are the ones that are currently broken on config_refactor branch any how (will be fixed before refactoring is complete)
